pkg/pop/authnscheme.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package pop
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// type of a PoP token, as opposed to "JWT" for a regular bearer token
+const popTokenType = "pop"
+
+// type representing the header of a PoP access token
+type header struct {
+	typ string
+	alg string
+	kid string
+}
+
+// returns a string representation of a header object
+func (h *header) ToString() string {
+	return fmt.Sprintf(`{"typ":"%s","alg":"%s","kid":"%s"}`, h.typ, h.alg, h.kid)
+}
+
+// returns a base-64 encoded string representation of a header object
+func (h *header) ToBase64() string {
+	return base64.RawURLEncoding.EncodeToString([]byte(h.ToString()))
+}
+
+// type representing the payload of a PoP token
+type payload struct {
+	at    string
+	ts    int64
+	host  string
+	jwk   string
+	nonce string
+}
+
+// returns a string representation of a payload object
+func (p *payload) ToString() string {
+	return fmt.Sprintf(`{"at":"%s","ts":%d,"u":"%s","cnf":{"jwk":%s},"nonce":"%s"}`, p.at, p.ts, p.host, p.jwk, p.nonce)
+}
+
+// returns a base-64 encoded representation of a payload object
+func (p *payload) ToBase64() string {
+	return base64.RawURLEncoding.EncodeToString([]byte(p.ToString()))
+}
+
+// type representing the signature of a PoP token
+type signature struct {
+	sig []byte
+}
+
+// returns a base-64 encoded representation of a signature object
+func (s *signature) ToBase64() string {
+	return base64.RawURLEncoding.EncodeToString(s.sig)
+}
+
+// type representing a PoP access token
+type PoPAccessToken struct {
```

Comment:
Do we need this to be exported ?

Reply:
Changed this to be unexported since it's not used outside this file